File: 卡码/代码随想录/Go/回溯/04-排列/LeetCode_0046.go
//46. 全排列
package main

func permute(nums []int) [][]int {
	res:=make([][]int,0)
	path:=make([]int,0)
	visited:=make([]bool,len(nums))
	for i:=0;i<len(nums);i++ {
		visited[i]=false
	}
	var backtracking func(nums []int)
	backtracking=func(nums []int){
		if len(path)==len(nums) {
			tmp:=make([]int,len(path))
			copy(tmp,path)
			res=append(res,tmp)
			return
		}
		for i:=0;i<len(nums);i++ {
			if visited[i]==true {
				continue
			}
			visited[i]=true
			path=append(path,nums[i])
			backtracking(nums)
			path=path[:len(path)-1]
			visited[i]=false
		}
	}
	backtracking(nums)
	return res
}

File: 卡码/代码随想录/Python/回溯/04-排列/LeetCode_0046.py
#46. 全排列
from typing import *

class Solution:
	def __init__(self):
		self.path:List[int]=[]
		self.res:List[List[int]]=[]
		self.visited:List[bool]=[]
	def backtracking(self,nums:List[int]):
		if len(self.path)==len(nums):
			self.res.append(self.path[:])
			return
		for i in range(len(nums)):
			if self.visited[i]:
				continue
			self.visited[i]=True
			self.path.append(nums[i])
			self.backtracking(nums)
			self.path.pop()
			self.visited[i]=False
	def permute(self, nums: List[int]) -> List[List[int]]:
		self.visited=[False]*len(nums)
		self.backtracking(nums)
		return self.res